k8s_auth: new k8s module for handling auth

mmazur opened the pull request:
##### SUMMARY
Some kubernetes distributions (like OpenShift) may require explicit login/logout steps before any of the `k8s*` modules can be used. This is the module that will handle those.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

modules/clustering/k8s/k8s_auth

mmazur commented:
No code yet, just the docs/options/return values/examples to hopefully get some early input.

One thing that bugs me in the example is that with this module one needs to keep repeating all the ssl settings and `host` urls in each `k8s*` module invocation. I think the proper way to address this would be to also modify the base k8s code so that all modules accept the `k8s_auth` facts dict and take all their configs from it. Something like this (as opposed to what's in the EXAMPLES section):

```
- k8s_facts:
    k8s_auth: {{ k8s_auth }}
    kind: Pod
  register: pod_list
```

Oh, and btw, I've kept all the option names. Though I'd really like to see `ssl_ca_cert` renamed to something like `ca_path` everywhere.

Ping @fabianvf

fabianvf commented:
Adding a new dict parameter that allows you to specify all your authentication options makes sense to me, @willthames any thoughts about that? Presumably we'd maintain the old parameters and just merge them into this dict in the code.

About `ssl_ca_cert`, do you know if it only works if you specify a path or can you provide one as a string? For some reason I thought it was either but I could be wrong.

willthames commented:
@fabianvf - I'd prefer to avoid adding special dicts that replace common settings - we seem to have enough issues getting auth params right without introducing further complexity (unless such a change came with a rock solid test suite)

Usually when passing common settings I suggest using YAML anchors: http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html#tasksmainyml

So here you'd do something like:

```
- k8s_auth:
     ...
   register: k8s_auth_result

- set_fact:
     &k8s_auth: "{{ k8s_auth_result }}"

- k8s_facts:
    <<: *k8s_auth
```

The downside of YAML anchors is they only work per file (but you could reuse `k8s_auth_result` in later files with the same `set_fact`)

In summary, I can see the argument for an `auth` parameter, and if an implementation came with a rigorous test suite, I'd support it, but YAML might suffice in the interim. And I think there's a mechanism used for GCE and AWS for reusing auth that might be better still, I'm just trying to find out what it's called, and whether we could use the same concept for k8s.

mmazur commented:
A combination of `module_defaults` for groups and `set_facts` does indeed look like the proper way to handle this. I will use a version of `set_facts` in the example for this module to make sure people end up using it.

@willthames are you 100% certain groups can't be named the same as an existing module? The implementation of groups for `module_defaults` expects one to use `group/` prefix, so I'm not entirely sure why it's a problem for there to exist both `group/k8s` and `k8s` as different arguments to `module_defaults`.

willthames commented:
@mmazur, it's not particularly well documented, as far as I can see, so I was making an assumption - using `group/k8s` would totally work for me. 

The nice thing is we don't have many modules so setting module_defaults per module isn't too much of a stretch until 2.8 is released.

mmazur committed:
k8s*: add a reference to k8s_auth in all the modules' descriptions

fabianvf commented:
For anyone else curious about defaults for module groups: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/user_guide/module_defaults_config.rst

I really like the module defaults. We should definitely include an example using it.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+    'host': {'required': True},
+    'username': {},
+    'password': {'no_log': True},
+    'ssl_ca_cert': {},
```
Should we `no_log` this as well? I still need to verify it can't be the actual cert content (rather than a path).

mmazur replied:
k8s/common.py says:

```
    'ssl_ca_cert': {
        'type': 'path',
    },
```
so it's definitively a path.

mmazur replied:
Hmm, as I look at the code, I'm not sure that's even implemented…

fabianvf replied:
It passes it through to the Kubernetes client's Configuration object which should handle it, but I don't think this path has been tested thoroughly. I'll see if I can find some time to investigate it soon

mmazur replied:
Ok. And btw, CA cert chains are just public keys, so there's no need to `no_log` them.

mmazur replied:
No, you wanted `no_log` to make sure there isn't hundreds of lines of pem files in the output. Makes sense.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+import traceback
+import requests
+from urllib3.util import make_headers
+from requests_oauthlib import OAuth2Session
```
Do we need to guard this import?

mmazur replied:
Yup, done.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+#    auth_endpoint = oauth_info['authorization_endpoint']
+#    token_endpoint = oauth_info['token_endpoint']
+    # TODO: remove DEBUG entries
+    auth_endpoint = f'{host}/oauth/authorize'
```
Are f-strings available in Python2?

mmazur replied:
This was only a thing I'm using for debugging to get the whole auth procedure through mitmproxy. I'll remove it anyways.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+        openshift_logout(host, api_key, ca_verify)
+
+    module.exit_json(
+        changed=False,
```
Maybe set this dict in the above if/else block

mmazur replied:
Good point, done.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+
+DOCUMENTATION = '''
+
+module: k8s_auth
```
It may make more sense to call it like `k8s_oauth_token` or something, I would assume I could use this for generic authorization otherwise. Alternatively we could just make it succeed with a no-op in the case that the cluster doesn't support oauth.

mmazur replied:
My thinking in naming it `k8s_auth` was that it should support all the auth mechanisms that will be used for k8s clusters now or in the future. With openshift that's only oauth, but if something somewhere started using zauth or ëauth, that support should be added to this module ideally with some autodetection (which should be trivial).

As to succeeding even when there's no oauth – that'd only be an option if the module first checked for oauth and if that was gone, tried to perform an operation with http basic auth to verify that that in fact works aka that the auth credentials are correct. (Same logic with using client cert instead of user+pwd, should someone add that eventually.) And depending on whether it got oauth or not, the return structure would either contain a token and an empty `password` field or an empty `api_key` field and valid user+pwd. So the fact/anchor creation code from the example would look like this:

```
- set_fact:
    k8s_auth_params: &k8s_auth_params
        host: "{{ k8s_auth_results.k8s_auth.host }}"
        ssl_ca_cert: "{{ k8s_auth_results.k8s_auth.ssl_ca_cert }}"
        verify_ssl: "{{ k8s_auth_results.k8s_auth.verify_ssl }}"
        api_key: "{{ k8s_auth_results.k8s_auth.api_key }}"
        username: "{{ k8s_auth_results.k8s_auth.username }}"
        password: "{{ k8s_auth_results.k8s_auth.password }}"
```

and the `k8s_auth_params` anchor could be used universally as it'd then work both with openshift and k8s clusters. Though again – that would depend on  `api_key` being non–null with openshift and `password` being non–null with http basic auth. I'm not exactly sure that doing the latter (returning a password and storing it) is a valid thing to do in ansible and whether the whole thing isn't too contrived.

@willthames, could you maybe comment as well?

fabianvf replied:
makes sense to me, I think in the long term it would be cool to be able to dump in all your parameters and get back a valid param block that you can add to module_defaults.

mmazur replied:
Ok, I've added a comment about this with a link to this particular part of the discussion, so that future changes stay in line with this general design.

As to implementing http basic auth – I've only gotten the go ahead to implement openshift for now (and the description for the module reflects that). If I get a green light from my team to take care of doing the group module defaults thing, I'll make sure to also add http basic auth fallback to this module and have this be as universal and easy to use as possible.

willthames replied:
By group module_defaults do you mean https://github.com/ansible/ansible/pull/51093 ?

mmazur replied:
Ah, great, having this exist will make my job much easier.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+    description:
+    - "Path to a CA certificate file used to verify connection to the API server. The full certificate chain
+      must be provided to avoid certificate validation errors."
+    - Only required if the CA certificate is not already present in the standard system CA trust store.
```
requests may not respect the standard CA trust store, so I think this might not be true

mmazur replied:
It does:

```
>>> import requests
>>> requests.get('https://github.com')
<Response [200]>
```

fabianvf replied:
I think requests uses the certs from the `certifi` package, not the system store. So if you have a custom certificate requests will not respect it.

mmazur replied:
Damn, this isn't the first time I've made this assumption. Thanks, I've removed the doc string.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,278 @@
+    - Provide a password for authenticating with the API server.
+  ssl_ca_cert:
+    description:
+    - "Path to a CA certificate file used to verify connection to the API server. The full certificate chain
```
I need to verify it has to be a path and that you can't specify the cert directly.

fabianvf reviewed (commented):
Looking good so far, have a couple questions/suggestions but I think (especially with module/group defaults) this will be a really nice feature.

willthames commented:
I've raised a PR to improve the module group defaults documentation (thanks for the pointer to that page @fabianvf )

https://github.com/ansible/ansible/pull/51045

mmazur committed:
k8s_auth: new k8s module for handling auth

mmazur commented:
Ok, I think I'm done with the coding. Everything I wanted to have working is working now. Removing the [WIP] tag.

Two things remaining:

- figuring out the name of this module
- if anybody has an idea what to do with the remaining CI errors, I'm all ears.

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+    verify_ssl:
+      description: "Whether or not to verify the API server's SSL certificates."
+      returned: success
+      type: bool
```
Needs a `default: true` to match the argument spec.

mmazur replied:
I've changed it, but it looks really weird to me…

mmazur replied:
Ok, please disregard that comment, fixed now.

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+
+        # python-requests takes either a bool or a path to a ca file as the 'verify' param
+        if verify_ssl and ssl_ca_cert:
+            self.con_verify_ca = ssl_ca_cert #path
```
comment should be `ssl_ca_cert  # path`

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+        if verify_ssl and ssl_ca_cert:
+            self.con_verify_ca = ssl_ca_cert #path
+        else:
+            self.con_verify_ca = verify_ssl #bool
```
two spaces before comment, one space after `#` - `verify_ssl  # bool`

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+            self.fail_json(msg="Something went wrong discovering OpenShift OAuth details.",
+                           exception=traceback.format_exc())
+
+
```
remove this blank line

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+        # In here we have `code` and `state`, I think `code` is the important one
+        qwargs = {}
+        for k, v in parse_qs(urlparse(ret.headers['Location']).query).items():
+            qwargs[k] = v[0] 
```
remove the space at the end of this line

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+    def fail_request(self, msg, **kwargs):
+        req_info = {}
+        for k, v in kwargs.items():
+            req_info['req_'+k] = v
```
should be `'req' + k`

willthames commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+
+# 3rd party imports
+try:
+    import requests
```
use ansible.module_utils.url rather than requests

mmazur replied:
Are there any docs for it?

mmazur replied:
Nope, and it doesn't support `ca_path`.

willthames reviewed (changes requested):
I've added notes for all the bits that are currently breaking CI. Let me know if any remain unclear.

mmazur commented:
Ok, I've checked `ansible.module_utils.urls.open_url()`, but it doesn't support providing one's own `ca_path`, so there's really no way for me to do a rewrite to remove `requests` dependency.

So, this is my final version (unless you want to continue the renaming discussion first) and please ignore that one CI error. ping @fabianvf @willthames

machacekondra commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+
+  - name: Preserve auth info as both a fact and a yaml anchor for easy access later
+    # Both the fact and the anchor are called 'k8s_auth_params'
+    set_fact:
```
This brings unnecessary complexity to user, why not make it simpler, and set such fact inside the module?

willthames replied:
Are you suggesting using `ansible_facts` for that (which seems like a bit of a hack?). I'm not aware of another way of transferring state from one module to other modules.

Even if using `ansible_facts` (and I don't think we should) we still need to document for people with `INJECT_FACTS_AS_VARS` set to False

mmazur replied:
I'm pretty sure Ondra's suggesting returning to my initial code with the module accepting an `k8s_auth` dict as a parameter (as well as returning one) and having all of the k8s modules accept that.

Summary of the discussion around this is – the proper way to do this kind of thing going forward will be with with [module defaults for groups](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/user_guide/playbooks_module_defaults.rst#module-defaults-groups), which became available since 2.7. So the yaml anchor usage in the EXAMPLE is basically just a stop–gap for how this will end up being used properly in the future.

So while it made sense to implement a dedicated auth dict parameter for groups of modules pre–2.7, it really would not be the proper implementation nowadays.

machacekondra replied:
Yes, I meant using `ansible_facts`, but now I realized you can't create anchor from that I guess.
Yeah the module defaults solve this quite nice, actually, so forget my comment.

machacekondra commented on lib/ansible/modules/clustering/k8s/k8s_auth.py:
```diff
@@ -0,0 +1,335 @@
+                           exception=traceback.format_exc())
+
+    def openshift_login(self):
+        os_oauth = OAuth2Session(client_id='openshift-challenging-client')
```
Actually, I think it would be better if such functionality is implemented inside openshift SDK, because more client codes could benefit from it, rather then other re-implement it their own way.

mmazur replied:
Well, the call by the openshift client maintainer (Fabian) was to do this directly in the module, at least initially. If we ever get a need to move that code into the lib, that can be a separate conversation.

fabianvf replied:
Yeah, I think it's best to do it here for now, but I could see incorporating this back into the client later.

mmazur committed:
k8s_auth: ignore E203

Can't use module_utils.urls, since that lacks user CA support, which is
a critical feature of what this module does.

mattclay commented:
@mmazur Docs fragments have moved, so this needs to be rebased.

mmazur commented:
@mattclay done.

fabianvf commented:
CI failures seem unrelated, can someone rekick this?

mattclay commented:
Failed tests have been restarted.

mmazur commented:
Still failing. Stuff like `"assertion": "current_time.stat.atime > previous_time.stat.atime"` suggests to me that the CI vms are wonky.

mmazur commented:
@mattclay can you run the tests again, maybe CI got stable(r) in the mean time.

fabianvf commented:
shipit

fabianvf commented:
bot_status

maxamillion commented:
rebuild_merge

mmazur commented:
ping

maxamillion commented:
rebuild_merge

mmazur commented:
I'm very confused on what's happening. Maybe this will work:

sudo_merge_force

fabianvf commented:
bot_status

fabianvf commented:
shipit

maxamillion commented:
!needs_revision

mmazur commented:
@mattclay k8s maintainers' shipits aren't enough here. Can you push this?
